### PR TITLE
Get LLVM smoke testing back on track after disabling enum->int casts

### PR DESCRIPTION
**Pull request by bradcray**
Summary: This should fix the llvm smoke-test that I broke...

In PR #10114, I stopped giving each enum an integer value since we're
moving towards supporting abstract enums that don't have underlying
integer values.  However, this broke LLVM code generation which seems
to generate enum values as integers (h/t to DavidK for sleuthing that
out for me).  This PR takes a stab at getting things running again,
and seems to be working, though I haven't taken the time to do a full
LLVM build + test on chapcs...

Specifically, what I've done here is preserve the old logic (which uses the
enums' initializer values when they're available) while having cases
that don't define initializers simply define the integer values to be
the enum's 0-based order.  This should match what the C back-end is
currently doing.

This seems to get testing working again, but seems as though it could
cause problems if/when there are conflicts between a user's enum values
and the ordinal values.  But, I suspect that the C back-end would have the
same issues.

I _suspect_ that the better thing to do would be to always generate enums
as ordinal values as proposed in issue #8561 for both the LLVM and C
back-ends.

### CODE REVIEW

**Comment by bradcray** [comment]
@mppf: I don't know how I turned a blind eye toward LLVM codegen in PR #10114 after wondering about it in some enum-based issues this week, but I did and broke smoke testing as a result...  As you'll read above, in this PR, I seem to have gotten testing working again, though I don't think the approach I took is principled (specifically, I wonder if we should/could always codegen enums as their ordinal values in the LLVM back-end, which would be similar to what I've most recently proposed in issue #8561 and remove the chances of duplicate values between the ordinal values and the user values that it seems this patch could lead to).

Anyway, if you could give this a post-merge review and continue this discussion with me, I'd appreciate it.

**Comment by bradcray** [comment]
After typing this and waiting for Travis to finish, I did a quick check to see if I could simply use the ordinal values (and never use the user's initial values), but this didn't work.  :(  This suggests that the LLVM-back-end is still using the user-provided integers for the enums in some cases (and maybe the C back-end is as well since I haven't changed that aspect of it).

Next, I'll try to write a test that exhibits a case where I'd expect this PR to cause problems...

File: compiler/codegen/type.cpp [hunk 1]
```diff
@@ -110,19 +110,24 @@ void EnumType::codegenDef() {
...
+      int order = 0;
       for_enums(constant, this) {
         //llvm::Constant *initConstant;
 
-        if(constant->init == NULL) INT_FATAL(this, "no constant->init");
-
-        VarSymbol* s = toVarSymbol(toSymExpr(constant->init)->symbol());
+        VarSymbol* s;
+        if (constant->init) {
+          s = toVarSymbol(toSymExpr(constant->init)->symbol());
```
Comment (dmk42) [inline compiler/codegen/type.cpp]: Would you mind adding a comment that all enum constants after the first explicitly-assigned one will have consant->init set?

**Comment by mppf** [comment]
@bradcray and @dmk42 - thanks for sorting it out. I agree that we should try to push towards enums generating only with ordinal values. One thing I don't understand is why we bother generating enum types differently than integer types. That is, the Chapel compiler knows each enum constant at compile-time, so why can't it just put in an integer value for them? Does the C/LLVM IR even need to know that the enum type != the appropriate integer type to store the ordinals?

**Comment by bradcray** [comment]
At first I thought when you said "enum constant" you meant the logical values that the user can associate with the enums and wrote this:

> the Chapel compiler knows each enum constant at compile-time

It doesn't in the current world in which an enum with no associated constants doesn't result in associated integer values anymore.  And if we relaxed the requirement that enum constants be params (issue #10115) it'd know less.

> so why can't it just put in an integer value for them?

If we continue to permit enums to have duplicate integer values (issue #9786) then replacing enums with integers results in problems due to the potential many-to-one nature.  E.g., mapping an execution-time enum to its corresponding string value causes problems.  This is another advantage of using ordinal representation—it's 1:1 by nature.

Then your last sentence made me think that you might mean "ordinal value" by "enum constant" in which case I think you're right—that the execution time representation of an enum would be no different than some integer value and that the Chapel compiler would take care of resolving and implementing all the operations that were enum-specific.

**Pull request merged**